hexl/fft/fwd-fft-avx512.cpp

```diff
@@ -0,0 +1,444 @@
+
+    __m512d v_X_real;
+    __m512d v_X_imag;
+    __m512d v_Y_real;
+    __m512d v_Y_imag;
+
+    ComplexLoadFwdInterleavedT2(X_real, &v_X_real, &v_Y_real);
+    ComplexLoadFwdInterleavedT2(X_imag, &v_X_imag, &v_Y_imag);
+
+    // Weights
+    __m512d v_W_real = _mm512_set_pd(
+        W_1C_intrlvd[6], W_1C_intrlvd[6], W_1C_intrlvd[4], W_1C_intrlvd[4],
+        W_1C_intrlvd[2], W_1C_intrlvd[2], W_1C_intrlvd[0], W_1C_intrlvd[0]);
+    __m512d v_W_imag = _mm512_set_pd(
+        W_1C_intrlvd[7], W_1C_intrlvd[7], W_1C_intrlvd[5], W_1C_intrlvd[5],
+        W_1C_intrlvd[3], W_1C_intrlvd[3], W_1C_intrlvd[1], W_1C_intrlvd[1]);
+    W_1C_intrlvd += 8;
+
+    ComplexFwdButterfly(&v_X_real, &v_X_imag, &v_Y_real, &v_Y_imag, v_W_real,
+                        v_W_imag);
+
+    _mm512_storeu_pd(v_X_pt_real, v_X_real);
+    _mm512_storeu_pd(v_X_pt_imag, v_X_imag);
+    v_X_pt_real += 2;
+    v_X_pt_imag += 2;
+    _mm512_storeu_pd(v_X_pt_real, v_Y_real);
+    _mm512_storeu_pd(v_X_pt_imag, v_Y_imag);
+
+    offset += 32;
+  }
+}
+
+void ComplexFwdT4(double_t* operand_8C_intrlvd, const double_t* W_1C_intrlvd,
+                  uint64_t m) {
+  size_t offset = 0;
+
+  // 2 | m guaranteed by n >= 16
+  HEXL_LOOP_UNROLL_4
+  for (size_t i = 0; i < (m >> 1); i += 2) {
+    double_t* X_real = operand_8C_intrlvd + offset;
+    double_t* X_imag = operand_8C_intrlvd + 8 + offset;
+
+    __m512d* v_X_pt_real = reinterpret_cast<__m512d*>(X_real);
+    __m512d* v_X_pt_imag = reinterpret_cast<__m512d*>(X_imag);
+
+    __m512d v_X_real;
+    __m512d v_X_imag;
+    __m512d v_Y_real;
+    __m512d v_Y_imag;
+
+    ComplexLoadFwdInterleavedT4(X_real, &v_X_real, &v_Y_real);
+    ComplexLoadFwdInterleavedT4(X_imag, &v_X_imag, &v_Y_imag);
+
+    // Weights
+    // x =  (11, 10,  9,  8, 3, 2, 1, 0)
+    // y =  (15, 14, 13, 12, 7, 6, 5, 4)
+    __m512d v_W_real = _mm512_set_pd(
+        W_1C_intrlvd[2], W_1C_intrlvd[2], W_1C_intrlvd[2], W_1C_intrlvd[2],
+        W_1C_intrlvd[0], W_1C_intrlvd[0], W_1C_intrlvd[0], W_1C_intrlvd[0]);
+    __m512d v_W_imag = _mm512_set_pd(
+        W_1C_intrlvd[3], W_1C_intrlvd[3], W_1C_intrlvd[3], W_1C_intrlvd[3],
+        W_1C_intrlvd[1], W_1C_intrlvd[1], W_1C_intrlvd[1], W_1C_intrlvd[1]);
+
+    W_1C_intrlvd += 4;
+
+    ComplexFwdButterfly(&v_X_real, &v_X_imag, &v_Y_real, &v_Y_imag, v_W_real,
+                        v_W_imag);
+
+    _mm512_storeu_pd(v_X_pt_real, v_X_real);
+    _mm512_storeu_pd(v_X_pt_imag, v_X_imag);
+    v_X_pt_real += 2;
+    v_X_pt_imag += 2;
+    _mm512_storeu_pd(v_X_pt_real, v_Y_real);
+    _mm512_storeu_pd(v_X_pt_imag, v_Y_imag);
+
+    offset += 32;
+  }
+}
+
+void ComplexFwdT8(double_t* operand_8C_intrlvd, const double_t* W_1C_intrlvd,
+                  uint64_t gap, uint64_t m) {
+  size_t offset = 0;
+
+  HEXL_LOOP_UNROLL_4
+  for (size_t i = 0; i < (m >> 1); i++) {
+    // Referencing operand
+    double_t* X_real = operand_8C_intrlvd + offset;
+    double_t* X_imag = operand_8C_intrlvd + 8 + offset;
```

Comment:
This may make a bit clearer. Or you could have it as X_imag = (*X_real) + 8;
```suggestion
    double_t* X_imag = operand_8C_intrlvd + offset + 8;
```